### ISSUE TITLE
direct comparison of linear layer with this paper 

### ISSUE THREAD

**Issue by dorost1234** [pengyuLPY/Virtual-Fully-Connected-Layer#4]
Dear Author 
I want to run a benchmarking to compare the two methods in terms of memory/time, here is my code for benchmarking:

```
import torch
import time
in_channels = 256
out_channels = 512
batch_size = 512
seq_length = 50
device = "cuda:0"
x = torch.randn((batch_size, seq_length, in_channels), requires_grad=True).to(device)
x = x.to(device)
standard_linear = torch.nn.Linear(in_channels, out_channels, bias=True).cuda()

def benchmark_module(module, iterations):
    start = time.time()
    for i in range(iterations):
        y = module(x)
        l = y.sum()
        l.backward()
    print(f"Execution: {((time.time()-start) / iterations):.6f}s.")

benchmark_module(module=standard_linear, iterations=1000)
peak_memory = (torch.cuda.max_memory_allocated() / 1024 ** 2)/1000
print("Memory utilization", peak_memory, "GB")
```

I need now to create a linear layer with your work and pass it to the function of "becnhmark_module", shall I create it as:
```
DictArcMarginProduct( in_features=128, out_features=200, out_features_test=74974, s=32.0, m=0.50, label_dict=None, easy_margin=False)
```
could you tell me please what are the following parameters:
```
out_features_test
s
m
label_dict
easy_margin
```
and how one can set these parameters to compare with a linear layer?
thanks

**Comment by pengyuLPY** [comment]
s, m and easy_margin: they are the hyper-parameters for ArcMargin Softmax loss function (arcFace loss function). 
label_dict: It is the rule to map the N identities (label belongs to R^N) to M groups (label belongs to R^M) . 
out_features_test: Because the free anchors would be 'borrowed' in our re-group strategy. The parameters of borrowed free anchors need to be recovered in the end of each iteration. We allocate a buffer to store the parameters. The size of buffer is in_features*out_features_test.  In our implementation, we set out_features_test equals to out_features simply (Line200 in Virtual-Fully-Connected-Layer/CNNs/resnet_identityMapping_face.py).

**Comment by dorost1234** [comment]
Hi
thanks for the response, I am sorry I am confused, if the size is in_dim*out_dim then this is taking as much memory/time as linear or perhaps more time since there are more operations, could you kindly let me know how time/memory is compared to a normal linear layer compared to your method? thanks

**Comment by pengyuLPY** [comment]
I think it is not a good idea to compare our implementation in this repository with standard FC. Because we implement the layer with arcFace loss function for face recognition task. the margin in the arcFace would cost some time/memory. 
We think it is more fair to compare our DictArcMarginProduct (virtual-FC) class with ArcMarginProduct in the time/memory. If you'd like to compare it with standard fully connected layer, we recommend you to implement the layer based on our theory.

**Comment by dorost1234** [comment]
Hi
thanks but i do not think this can be even replaced by a fully connected
layer your work seems to be applicable only for softmax and my question
would be why this is not compared to hierchical softmax and more efficient
variants of it. thanks

**Comment by pengyuLPY** [comment]
I implement you code to compare our DictArcMarginProduct and ArcMarginProduct. The code is shown in the following:

import torch
import time
from CNNs.DictarcFace import DictArcMarginProduct
from CNNs.arcFace import ArcMarginProduct
in_channels = 256
out_channels = 512
batch_size = 512
iterations=100
label_dict = {i:i for i in range(out_channels)}
device = "cuda:0"
x = torch.randn((batch_size,  in_channels), requires_grad=True).to(device)
label = torch.ones((batch_size)).to(device)
x = x.to(device)
label = label.to(device)
label_set = set(label.cpu().tolist())


test_type = 'ArcMarginProduct'
#test_type = 'virtual_linear'


if test_type == 'ArcMarginProduct':
    def benchmark_module(module, iterations):
        start = time.time()
        for i in range(iterations):
            y = module(x, label)
            l = y.sum()
            l.backward()
        print(f"Execution: {((time.time()-start) / iterations):.6f}s.")


    #standard_linear = torch.nn.Linear(in_channels, out_channels, bias=True).cuda()
    standard_linear = ArcMarginProduct(in_channels, out_channels).cuda()
    print('ArcMarginProduct')
    benchmark_module(module=standard_linear, iterations=iterations)
    peak_memory = (torch.cuda.max_memory_allocated() / 1024 ** 2)/1000
    print("Memory utilization", peak_memory, "GB")


else:
    virtual_linear = DictArcMarginProduct(in_features=in_channels, out_features=out_channels, out_features_test=out_channels,  s=32.0, m=0.50, label_dict=label_dict, easy_margin=False).cuda()

    def benchmark_module_virtual_fc(module, iterations):
        start = time.time()
        for i in range(iterations):
            y,_ = module(x, label, label_set)
            l = y.sum()
            l.backward()
        print(f"Execution: {((time.time()-start) / iterations):.6f}s.")


    print('virtual_linear')
    benchmark_module_virtual_fc(module=virtual_linear, iterations=iterations)
    peak_memory = (torch.cuda.max_memory_allocated() / 1024 ** 2)/1000
    print("Memory utilization", peak_memory, "GB")


it print that:
ArcMarginProduct
Execution: 0.003019s.
Memory utilization 0.01201025390625 GB

virtual_linear
Execution: 0.003641s.
Memory utilization 0.0125146484375 GB

**Comment by pengyuLPY** [comment]
Sure, our virtual FC can not replace the standard FC at every situation. It can substitute the FC layer only when the FC layer is used as a linear classifier. In our experiments, we just implement our proposal with ArcMarginProduct because the ArcMarginProduct achieved the state-of-the-art performance in the face recognition fields.

By the way, sorry for the code format in the previous message.

**Comment by pengyuLPY** [comment]
import torch
import time
from CNNs.DictarcFace import DictArcMarginProduct
from CNNs.arcFace import ArcMarginProduct
in_channels = 256
out_channels = 512
batch_size = 512
iterations=100
label_dict = {i:i for i in range(out_channels)}
device = "cuda:0"
x = torch.randn((batch_size,  in_channels), requires_grad=True).to(device)
label = torch.ones((batch_size)).to(device)
x = x.to(device)
label = label.to(device)
label_set = set(label.cpu().tolist())


test_type = 'ArcMarginProduct'
#test_type = 'virtual_linear'


if test_type == 'ArcMarginProduct':
    def benchmark_module(module, iterations):
        start = time.time()
        for i in range(iterations):
            y = module(x, label)
            l = y.sum()
            l.backward()
        print(f"Execution: {((time.time()-start) / iterations):.6f}s.")


    #standard_linear = torch.nn.Linear(in_channels, out_channels, bias=True).cuda()
    standard_linear = ArcMarginProduct(in_channels, out_channels).cuda()
    print('ArcMarginProduct')
    benchmark_module(module=standard_linear, iterations=iterations)
    peak_memory = (torch.cuda.max_memory_allocated() / 1024 ** 2)/1000
    print("Memory utilization", peak_memory, "GB")


else:
    virtual_linear = DictArcMarginProduct(in_features=in_channels, out_features=out_channels, out_features_test=out_channels,  s=32.0, m=0.50, label_dict=label_dict, easy_margin=False).cuda()

    def benchmark_module_virtual_fc(module, iterations):
        start = time.time()
        for i in range(iterations):
            y,_ = module(x, label, label_set)
            l = y.sum()
            l.backward()
        print(f"Execution: {((time.time()-start) / iterations):.6f}s.")


    print('virtual_linear')
    benchmark_module_virtual_fc(module=virtual_linear, iterations=iterations)
    peak_memory = (torch.cuda.max_memory_allocated() / 1024 ** 2)/1000
    print("Memory utilization", peak_memory, "GB")

**Comment by pengyuLPY** [comment]
`import torch
import time
from CNNs.DictarcFace import DictArcMarginProduct
from CNNs.arcFace import ArcMarginProduct
in_channels = 256
out_channels = 512
batch_size = 512
iterations=100
label_dict = {i:i for i in range(out_channels)}
device = "cuda:0"
x = torch.randn((batch_size,  in_channels), requires_grad=True).to(device)
label = torch.ones((batch_size)).to(device)
x = x.to(device)
label = label.to(device)
label_set = set(label.cpu().tolist())


test_type = 'ArcMarginProduct'
#test_type = 'virtual_linear'


if test_type == 'ArcMarginProduct':
    def benchmark_module(module, iterations):
        start = time.time()
        for i in range(iterations):
            y = module(x, label)
            l = y.sum()
            l.backward()
        print(f"Execution: {((time.time()-start) / iterations):.6f}s.")


    #standard_linear = torch.nn.Linear(in_channels, out_channels, bias=True).cuda()
    standard_linear = ArcMarginProduct(in_channels, out_channels).cuda()
    print('ArcMarginProduct')
    benchmark_module(module=standard_linear, iterations=iterations)
    peak_memory = (torch.cuda.max_memory_allocated() / 1024 ** 2)/1000
    print("Memory utilization", peak_memory, "GB")


else:
    virtual_linear = DictArcMarginProduct(in_features=in_channels, out_features=out_channels, out_features_test=out_channels,  s=32.0, m=0.50, label_dict=label_dict, easy_margin=False).cuda()

    def benchmark_module_virtual_fc(module, iterations):
        start = time.time()
        for i in range(iterations):
            y,_ = module(x, label, label_set)
            l = y.sum()
            l.backward()
        print(f"Execution: {((time.time()-start) / iterations):.6f}s.")


    print('virtual_linear')
    benchmark_module_virtual_fc(module=virtual_linear, iterations=iterations)
    peak_memory = (torch.cuda.max_memory_allocated() / 1024 ** 2)/1000
    print("Memory utilization", peak_memory, "GB")`

**Comment by pengyuLPY** [comment]
`import torch
import time
from CNNs.DictarcFace import DictArcMarginProduct
from CNNs.arcFace import ArcMarginProduct
in_channels = 256
out_channels = 512
batch_size = 512
iterations=100
label_dict = {i:i for i in range(out_channels)}
device = "cuda:0"
x = torch.randn((batch_size,  in_channels), requires_grad=True).to(device)
label = torch.ones((batch_size)).to(device)
x = x.to(device)
label = label.to(device)
label_set = set(label.cpu().tolist())


test_type = "ArcMarginProduct"
#test_type = "virtual_linear"


if test_type == "ArcMarginProduct":
    def benchmark_module(module, iterations):
        start = time.time()
        for i in range(iterations):
            y = module(x, label)
            l = y.sum()
            l.backward()
        print(f"Execution: {((time.time()-start) / iterations):.6f}s.")


    #standard_linear = torch.nn.Linear(in_channels, out_channels, bias=True).cuda()
    standard_linear = ArcMarginProduct(in_channels, out_channels).cuda()
    print("ArcMarginProduct")
    benchmark_module(module=standard_linear, iterations=iterations)
    peak_memory = (torch.cuda.max_memory_allocated() / 1024 ** 2)/1000
    print("Memory utilization", peak_memory, "GB")


else:
    virtual_linear = DictArcMarginProduct(in_features=in_channels, out_features=out_channels, out_features_test=out_channels,  s=32.0, m=0.50, label_dict=label_dict, easy_margin=False).cuda()

    def benchmark_module_virtual_fc(module, iterations):
        start = time.time()
        for i in range(iterations):
            y,_ = module(x, label, label_set)
            l = y.sum()
            l.backward()
        print(f"Execution: {((time.time()-start) / iterations):.6f}s.")


    print("virtual_linear")
    benchmark_module_virtual_fc(module=virtual_linear, iterations=iterations)
    peak_memory = (torch.cuda.max_memory_allocated() / 1024 ** 2)/1000
    print("Memory utilization", peak_memory, "GB")`

**Comment by pengyuLPY** [comment]
```
import torch
import time
from CNNs.DictarcFace import DictArcMarginProduct
from CNNs.arcFace import ArcMarginProduct
in_channels = 256
out_channels = 512
batch_size = 512
iterations=100
label_dict = {i:i for i in range(out_channels)}
device = "cuda:0"
x = torch.randn((batch_size,  in_channels), requires_grad=True).to(device)
label = torch.ones((batch_size)).to(device)
x = x.to(device)
label = label.to(device)
label_set = set(label.cpu().tolist())


test_type = "ArcMarginProduct"
#test_type = "virtual_linear"


if test_type == "ArcMarginProduct":
    def benchmark_module(module, iterations):
        start = time.time()
        for i in range(iterations):
            y = module(x, label)
            l = y.sum()
            l.backward()
        print(f"Execution: {((time.time()-start) / iterations):.6f}s.")


    #standard_linear = torch.nn.Linear(in_channels, out_channels, bias=True).cuda()
    standard_linear = ArcMarginProduct(in_channels, out_channels).cuda()
    print("ArcMarginProduct")
    benchmark_module(module=standard_linear, iterations=iterations)
    peak_memory = (torch.cuda.max_memory_allocated() / 1024 ** 2)/1000
    print("Memory utilization", peak_memory, "GB")


else:
    virtual_linear = DictArcMarginProduct(in_features=in_channels, out_features=out_channels, out_features_test=out_channels,  s=32.0, m=0.50, label_dict=label_dict, easy_margin=False).cuda()

    def benchmark_module_virtual_fc(module, iterations):
        start = time.time()
        for i in range(iterations):
            y,_ = module(x, label, label_set)
            l = y.sum()
            l.backward()
        print(f"Execution: {((time.time()-start) / iterations):.6f}s.")


    print("virtual_linear")
    benchmark_module_virtual_fc(module=virtual_linear, iterations=iterations)
    peak_memory = (torch.cuda.max_memory_allocated() / 1024 ** 2)/1000
    print("Memory utilization", peak_memory, "GB")
```

**Comment by khawar-islam** [comment]
@dorost1234 Would it be possible to replace nn.Linear to author layer? If yes, please mention some little example.
